src/main/java/net/dv8tion/jda/api/exceptions/InsufficientPermissionException.java

```diff
@@ -16,17 +16,55 @@
 
 package net.dv8tion.jda.api.exceptions;
 
+import net.dv8tion.jda.api.JDA;
 import net.dv8tion.jda.api.Permission;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.GuildChannel;
+import net.dv8tion.jda.internal.utils.Checks;
 
 public class InsufficientPermissionException extends PermissionException
 {
-    public InsufficientPermissionException(Permission permission)
+    private final long guildId;
+    private final long channelId;
+
+    public InsufficientPermissionException(Guild guild, GuildChannel channel, Permission permission)
     {
         super(permission, "Cannot perform action due to a lack of Permission. Missing permission: " + permission.toString());
+        this.guildId = guild.getIdLong();
+        this.channelId = channel == null ? 0 : channel.getIdLong();
     }
 
-    public InsufficientPermissionException(Permission permission, String reason)
+    public InsufficientPermissionException(Guild guild, GuildChannel channel, Permission permission, String reason)
```

Comment:
Permission should come first in these constructor as it is the more important bit.

Reply:
We should also separate the constructors out to add clarity to the calling code.
```
InsufficientPermissionException(Permission, Guild)
InsufficientPermissionException(Permission, Guild, String)
InsufficientPermissionException(Permission, GuildChannel)
InsufficientPermissionException(Permission, GuildChannel, String)
```

This will prevent us from having to pass in nulls or pass additional values when they can be retrieved via other passed in values.
Example: pre-retrieved guild off of text channel `text`
```
new InsufficientPermissionException(Permission, text.getGuild(), text)
```